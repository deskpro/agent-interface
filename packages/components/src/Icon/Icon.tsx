import * as React from "react";
import "@deskpro/agent-interface-style/dist/elements/dp-Icons.css";
import icons from "./icons";

export type IconProps = {
  name: string;
  size?: number;
  color?: "default" | "primary" | "dark";
};

const colors = {
  default: "#979797",
  primary: "#3a8dde",
  dark: "#8B9293"
};

const Icon: React.SFC<IconProps> = ({ name, size = 24, color = "default" }) => {
  const SvgIcon = icons[name];
  if (!SvgIcon) {
    return null;
  }
  return (
    <span className={`dp-Icon dp-Icon--${name}`}>
      <SvgIcon
        fill={colors[color] || colors.default}
        width={size}
        height={size}
      />
    </span>
  );
};

Icon.defaultProps = {
  color: "default",
  size: 24
};

export default Icon;
