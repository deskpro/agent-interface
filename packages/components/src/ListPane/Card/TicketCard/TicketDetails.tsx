import * as React from "react";
import UserInfo, { UserInfoProps } from "../../../User/UserInfo/UserInfo";
import TagBar from "../../../elements/Badges/TagBar";
import Tag from "../../../elements/Badges/Tag";
import Meta from "../../../elements/Meta/Meta";

export type TicketDetailsProps = {
  user: UserInfoProps;
  labels?: Array<{ name: string; color?: DPColor }>;
  properties?: Array<{ name: string; value: string }>;
};

const TicketDetails: React.FC<TicketDetailsProps> = ({
  user,
  labels = [],
  properties = []
}) => (
  <>
    <UserInfo {...user} />
    {!!labels.length && (
      <TagBar>
        {labels.map(({ name, color }) => (
          <Tag key={name} color={color}>
            {name}
          </Tag>
        ))}
      </TagBar>
    )}
    {!!properties.length && (
      <span className="dp-JustifyStart">
        {properties.map(prop => (
          <Meta key={prop.name} {...prop} />
        ))}
      </span>
    )}
  </>
);

export default TicketDetails;
