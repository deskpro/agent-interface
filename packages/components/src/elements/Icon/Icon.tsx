import * as React from "react";
import classNames from "classnames";

import "@deskpro/agent-interface-style/dist/elements/dp-Icons.css";
import icons from "./icons";

export type IconProps = React.ComponentProps<"span"> & {
  name: string;
  size?: number;
  circle?: boolean;
  color?: DPColor;
  iconRef?: React.Ref<HTMLSpanElement>;
  onClick?(
    e: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>
  ): void;
};

const Icon: React.FC<IconProps> = ({
  name,
  size = 24,
  color,
  circle,
  iconRef,
  onClick,
  ...props
}) => {
  const SvgIcon = icons[name];
  if (!SvgIcon) {
    return null;
  }
  return (
    <span
      className={classNames(`dp-Icon Icon--${name}`, {
        "Icon--circle": circle,
        [`Icon--${color}`]: !!color
      })}
      role="button"
      tabIndex={-1}
      onClick={onClick}
      onKeyDown={onClick}
      ref={iconRef}
      {...props}
    >
      <SvgIcon width={size} height={size} />
    </span>
  );
};

Icon.defaultProps = {
  color: undefined,
  size: 24
};

export default React.memo(
  React.forwardRef<HTMLSpanElement, IconProps>((props, ref) => (
    <Icon {...props} iconRef={ref as React.Ref<HTMLSpanElement>} />
  ))
);
