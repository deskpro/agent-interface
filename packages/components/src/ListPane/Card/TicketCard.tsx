import * as React from "react";

import Card, { CheckableCardProps, CardProps, ICardModel } from "./Card";
import CardTitle from "./CardTitle";
import Checkbox from "../../inputs/Checkbox/Checkbox";
import IconGroup from "../../elements/Icon/IconGroup";
import Icon from "../../elements/Icon/Icon";

export interface TicketModel extends ICardModel {
  title: string;
  isFavorite?: boolean;
}

export type TicketCardProps = {
  icons?: React.ReactNode;
  statusLabel?: React.ReactNode;
  onFavoriteToggle?(e: React.MouseEvent<HTMLElement>): void;
};

const TicketCard: React.FC<
  CardProps<TicketModel> & CheckableCardProps & TicketCardProps
> = ({
  title,
  model,
  icons,
  statusLabel,
  checkable = false,
  checked,
  onCheck,
  onFavoriteToggle,
  renderBody,
  ...props
}) => (
  <Card model={model} {...props}>
    <span className="dp-LevelCol">
      {!!checkable && (
        <Checkbox
          id={model.id as string}
          value={model.id}
          checked={checked}
          onChange={onCheck}
        />
      )}
    </span>
    <span className="dp-LevelCol">
      <CardTitle
        title={
          <>
            {model.title}
            {model.isFavorite && (
              <Icon
                name="star"
                size={15}
                color="primary"
                onClick={onFavoriteToggle}
              />
            )}
          </>
        }
      />
      {!!renderBody && renderBody(model)}
    </span>
    <span className="dp-LevelCol dp-LevelRight">
      {!!icons && <IconGroup>{icons}</IconGroup>}
      {!!statusLabel && <span className="dp-TimeStatus">{statusLabel}</span>}
    </span>
  </Card>
);

export default TicketCard;
