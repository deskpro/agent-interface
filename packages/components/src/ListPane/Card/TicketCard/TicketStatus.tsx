import * as React from "react";
import IconGroup from "../../../elements/Icon/IconGroup";
import Icon from "../../../elements/Icon/Icon";
import Tag from "../../../elements/Badges/Tag";

export type TicketStatusProps = {
  isLocked?: boolean;
  icons: React.ReactNode;
  lastUpdate?: string | Date;
  nextSlaStatus?: {
    status: "ok" | "warning" | "fail";
    isComplete?: boolean;
    nextTriggerDate: string | Date;
  };
};

const TicketStatus: React.FC<TicketStatusProps> = ({
  isLocked = false,
  icons,
  lastUpdate,
  nextSlaStatus
}) => (
  <>
    {(!!icons || isLocked) && (
      <IconGroup>
        {!!isLocked && <Icon name="lock" size={18} color="grey-darkest" />}
        {icons}
      </IconGroup>
    )}
    {(!!lastUpdate || !!nextSlaStatus) && (
      <span className="dp-TimeStatus">
        {!!nextSlaStatus && !nextSlaStatus.isComplete ? (
          <>
            {nextSlaStatus.status === "ok" && (
              <Tag size="small" color="grey">
                {nextSlaStatus.nextTriggerDate}
              </Tag>
            )}
            {nextSlaStatus.status === "warning" && (
              <Tag size="small" color="purple">
                {nextSlaStatus.nextTriggerDate}
              </Tag>
            )}
            {nextSlaStatus.status === "fail" && (
              <Tag size="small" color="danger">
                Failed
              </Tag>
            )}
          </>
        ) : (
          lastUpdate
        )}
      </span>
    )}
  </>
);

export default TicketStatus;
