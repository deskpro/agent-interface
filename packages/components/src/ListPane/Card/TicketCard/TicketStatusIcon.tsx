import * as React from "react";
import Icon from "../../../elements/Icon/Icon";
import Badge from "../../../elements/Badges/Badge";

export type TicketStatusIconProps = {
  status: string;
  urgency?: number;
};

const badgeColors: DPColor[] = [
  "grey",
  "neutral",
  "neutral",
  "neutral",
  "warning",
  "warning",
  "warning",
  "danger",
  "danger",
  "danger"
];

const TicketStatusIcon: React.FC<TicketStatusIconProps> = ({
  status,
  urgency
}) => {
  const [mainStatus] = status.split(".");

  if (mainStatus === "awaiting_agent") {
    return (
      <Badge
        type="round"
        color={
          !!urgency && urgency > 0 && urgency <= badgeColors.length
            ? badgeColors[urgency - 1]
            : "grey"
        }
      >
        {urgency}
      </Badge>
    );
  }
  switch (status) {
    case "awaiting_user":
      return <Icon name="avatar" size={18} color="pink" title={status} />;

    case "pending":
      return <Icon name="history" size={18} color="purple" title={status} />;

    case "resolved":
      return <Icon name="approve" size={18} color="success" title={status} />;

    case "hidden":
      return <Icon name="on-hold" size={18} color="grey" title={status} />;

    default:
      return null;
  }
};

export default TicketStatusIcon;
