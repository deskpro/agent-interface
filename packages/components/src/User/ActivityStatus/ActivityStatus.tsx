import * as React from "react";
import classNames from "classnames";

import Icon from "../../elements/Icon/Icon";

import "@deskpro/agent-interface-style/dist/components/dp-icons.css";

export type ActivityStatusProps = {
  status: "writing" | "viewing";
  color: DPColor;
  className?: string;
};

const ActivityStatus: React.SFC<ActivityStatusProps> = ({
  status,
  color,
  children,
  className
}) => (
  <span
    className={classNames(
      "dp-ActivityStatus",
      `ActivityStatus--${status}`,
      className,
      { [`ActivityStatus--${color}`]: !!color }
    )}
  >
    <Icon
      name={status === "writing" ? "edit" : "view"}
      color={color}
      size={15}
    />
    {children}
  </span>
);

export default ActivityStatus;
