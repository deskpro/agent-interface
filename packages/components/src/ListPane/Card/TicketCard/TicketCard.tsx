import * as React from "react";

import SimpleCard, { SimpleCardProps } from "../SimpleCard";
import Icon from "../../../elements/Icon/Icon";
import TicketStatus from "./TicketStatus";
import TicketStatusIcon from "./TicketStatusIcon";
import TicketDetails from "./TicketDetails";

type TicketCardSubcomponents = {
  Status: typeof TicketStatus;
  StatusIcon: typeof TicketStatusIcon;
  Details: typeof TicketDetails;
};

export type TicketCardProps = {
  isFavorite?: boolean;
  onFavoriteToggle?(cardId: React.Key, e: React.MouseEvent<HTMLElement>): void;
};

const TicketCard: React.FC<SimpleCardProps & TicketCardProps> &
  TicketCardSubcomponents = ({
  cardId,
  title,
  isFavorite = false,
  onFavoriteToggle,
  ...props
}) => (
  <SimpleCard
    cardId={cardId}
    {...props}
    title={
      <>
        {title}
        {isFavorite && (
          <Icon
            name="star"
            size={15}
            color="primary"
            onClick={(e: React.MouseEvent<HTMLElement>) =>
              onFavoriteToggle && onFavoriteToggle(cardId, e)
            }
          />
        )}
      </>
    }
  />
);

TicketCard.Status = TicketStatus;
TicketCard.StatusIcon = TicketStatusIcon;
TicketCard.Details = TicketDetails;

export default TicketCard;
