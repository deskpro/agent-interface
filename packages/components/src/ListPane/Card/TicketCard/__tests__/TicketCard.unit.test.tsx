import * as React from "react";
import { render } from "react-testing-library";
import { addHours, addMinutes, subMinutes } from "date-fns";

import ActivityStatus from "../../../../User/ActivityStatus/ActivityStatus";
import Icon from "../../../../elements/Icon/Icon";
import TicketCard from "../TicketCard";

describe("<TicketCard />", () => {
  it("it should match its snapshot", () => {
    const date = new Date("2019-02-02 15:55:18");
    const onCheck = jest.fn();
    const { container } = render(
      <TicketCard
        className="test-card"
        cardId={1}
        title="Card title"
        checkable
        onCheck={onCheck}
        icons={
          <>
            <ActivityStatus status="writing">
              <div>Avatar</div>
            </ActivityStatus>
            <Icon name="beetle" circle />
            <TicketCard.StatusIcon
              urgency={1}
              status="awaiting_agent.announcement"
            />
          </>
        }
        status={
          <TicketCard.Status
            lastUpdate={subMinutes(date, 22)}
            nextSlaStatus={{
              isComplete: false,
              nextTriggerDate: addHours(addMinutes(date, 14), 2),
              status: "ok"
            }}
          />
        }
      >
        <TicketCard.Details
          user={{
            name: "John Doe",
            email: "john.doe@example.com",
            avatar: true
          }}
        />
      </TicketCard>
    );
    expect(container.querySelector(".test-card")).toMatchSnapshot();
  });
});
