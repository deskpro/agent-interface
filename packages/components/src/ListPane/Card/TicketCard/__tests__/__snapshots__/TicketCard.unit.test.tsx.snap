// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicketCard /> should match its snapshot 1`] = `
<div
  class="dp-Card test-card"
  role="link"
  tabindex="-1"
>
  <div
    class="dp-LevelItem Level-justify-between"
  >
    <span
      class="dp-LevelLeft"
    >
      <span
        class="dp-CustomCheckbox dp-LevelTop"
      >
        <input
          class="dp-ControlInput"
          id="1"
          type="checkbox"
          value="1"
        />
        <label
          class=""
          for="1"
        />
      </span>
      <span
        class="dp-Card-title"
      >
        Card title
      </span>
    </span>
    <span
      class="dp-LevelRight"
    >
      <span
        class="dp-IconGroup"
      >
        <span
          class="dp-ActivityStatus ActivityStatus--writing ActivityStatus--primary"
        >
          <div>
            Avatar
          </div>
        </span>
        <span
          class="dp-Badge Badge--round Badge--grey"
        >
          
          1
        </span>
      </span>
    </span>
  </div>
  <div
    class="dp-LevelItem Level-justify-between"
  >
    <span
      class="dp-LevelLeft"
    >
      <span
        class="dp-UserInfo"
      >
        <span
          class="dp-UserName"
        >
          John Doe
          <a
            class="dp-UserEmail"
            href="mailto:john.doe@example.com"
          >
            &lt;
            john.doe@example.com
            &gt;
          </a>
        </span>
      </span>
    </span>
    <span
      class="dp-LevelRight"
    >
      <span
        class="dp-TimeStatus"
      >
        <span
          class="dp-Tag Tag--grey Tag--small"
        >
          2 months
        </span>
      </span>
    </span>
  </div>
</div>
`;
